Additions and revisions for v0.2.0 release

sgreenbury opened the pull request:
Closes #58.

This adds:
- Fix conflicting help messages
- BBox for `Data` subcommand
- `--no-geometry` option to exclude geometries from request
- Impl of `countries` subcommand
- Quickstart
- Adds quiet mode without progress bar/spinner to optionally enable logs to readable

sgreenbury commented on src/cli.rs:
```diff
@@ -107,6 +138,10 @@ impl RunCommand for DataCommand {
         let search_results = popgetter.search(self.search_params_args.clone().into());
 
+        // Make DataRequestSpec
+        // TODO: possibly implement From<(DataCommand, SearchParams)> for DataRequestSpec
```
```suggestion
        // TODO: consider alternative `From` impls as part of #67
```

sgreenbury commented on src/cli.rs:
```diff
@@ -153,13 +189,16 @@ impl RunCommand for DataCommand {
-        help = "Bounding box in which to get the results"
-    )]
-    bbox: Option<BBox>,
+    // TODO: consider implementation of bbox as part of:
```
```suggestion
    // TODO: consider implementation of bbox for metrics command as part of:
```

sgreenbury commented on src/search.rs:
```diff
@@ -353,21 +369,41 @@ impl SearchResults {
         // TODO Handle multiple responses
         if all_geom_files.len() > 1 {
-            panic!("Multiple geometries not yet supported");
+            todo!("Multiple geometries not yet supported");
```
```suggestion
            unimplemented!("Multiple geometries not yet supported");
```

sgreenbury commented on src/search.rs:
```diff
@@ -353,21 +369,41 @@ impl SearchResults {
+                    data_request_spec.region
+                );
+            }
+            debug!("{:#?}", data_request_spec.region.first().unwrap().bbox());
```
```suggestion

```

andrewphilipsmith reviewed (approved):
Overall this looks great and is a nice improvement on the CLI.

My only concern for v0.2 is the `--bbox` option and how it handles non Long/Lat values. A good enough solution for v0.2 might be to simply add a caveat to the --help description and maybe also a warning to the results (along the lines of "some results may have been excluded because...").

Any other comments are firmly in the "for a later date" category.

# Bounding Box - non Long/Lat values

Do we know if this works if the geometry data is not (currently) held in a Long/Lat? Eg the England and Wales geometries are all in British National Grid? Is there any reprojection of these? If not is there a risk that this could cause more confusion that it solves?

# Bounding Box - also for `metrics` subcommand?

Possibly an enchancement for >v0.2: Would it be useful to have `--bbox` option for the metrics subcommand?

# `--no-geometry` option

Not for v0.2, but some enhancements for future versions:

- Subjective, but in general, I prefer parameters to be specified in the positive (eg `geometry=TRUE` is better than `no-geometry=FALSE`). Having mutually exclusive options `--no-geometry` and `--inc-geometry` also avoids double-negatives

Ideally the default behaviour for this parameter should follow the output format selected. The combination of `--output-format=geojson --no-geometry` is equivalent to `--output-format=json`. In which case the defaults could be:

| output-format | include geometry by default |
| - | - |
| csv | No |
| json | No |
| geojson | Yes |
| geojsonseq | Yes |

# Countries subcommand

Very handy!

Nit-picking and not for v0.2: We might want to find a more generic alias for this command (and table column headings) that acknowledges that the coverage of some datasets might not align with country boundaries (the England and Wales census being the current example).

sgreenbury commented on README.md:
```diff
@@ -2,17 +2,27 @@
-
+| popgetter | popgetter-cli |
+| --------- | ------------- |
+| 0.1.0     | 0.1.0         |
```
Since there isn't a v0.1.0 of the CLI, perhaps this would be better:
```suggestion
| 0.1.0     |          |
| 0.2.0     | 0.2.0         |
```

penelopeysm commented on src/config.rs:
```diff
@@ -9,9 +9,7 @@ pub struct Config {
-            base_path:
-                "https://popgetter.blob.core.windows.net/popgetter-dagster-test/test_v2_release"
-                    .into(),
+            base_path: "https://popgetter.blob.core.windows.net/releases/v0.2.0".into(),
```
I posted a long thing about versioning in Slack, but I think the only bit that needs to be decided now is whether we want to drop the patch version from the Azure path, i.e. use `v0.2` instead of `v0.2.0`. My rationale is that this means we don't need to republish data for non-breaking changes.

sgreenbury replied:
Thanks for the comments and suggestions on this @penelopeysm. This sounds good to me and simplifies the extent of the data publishing requirement and guides the versioning process. I've updated in the config to `v0.2` for now and added a todo regarding automating it as you mentioned (perhaps with [built](https://docs.rs/built/latest/built/)).

penelopeysm commented on README.md:
```diff
@@ -2,17 +2,28 @@
+| 0.2.0     | 0.2.0         |
+| ...       | ...           |
 
 ## Developer notes
 
```
I think we can remove this section on developer notes, it's no longer relevant (we aren't using the exported schema).

The section above (with the version tables) _may_ need to be updated depending on what we decide to do with versioning; however, what I proposed in Slack doesn't invalidate the above (it'll still be true that 0.2.0 corresponds to 0.2.0) so we can leave this info for now.

sgreenbury replied:
Thanks for mentioning this - I've now removed this.

penelopeysm reviewed (commented):
Sorry for being a busybody. Just a couple of small notes1

sgreenbury commented:
Thanks @andrewphilipsmith and @penelopeysm for the reviews and comments!

To address the `BBox` issue @andrewphilipsmith raises, wehave now:
- Tested that the `BBox` functions with respect to the coordinate system of the geometry requested and updated the the type in the `--help` message to: `LEFT,BOTTOM,RIGHT,TOP` so that it is independent of coordinate system
- Identified the geometries for each country currently included and updated the `--help` message to include details of the coordinate reference system used in each cases.
- Added a `warn!()` with message when the bbox is used to filter the data and enabled `RUST_LOG=warn` to be default when the CLI runs so that this is flagged by default
- Added a quiet mode to disable the progress indicator when logging is enabled

I'll go ahead and merge this now and open new issues to follow up on automated path for versions, flags and bbox for metrics.
